crates/task-impls/src/consensus/mod.rs

```diff
@@ -143,6 +145,182 @@ pub struct ConsensusTaskState<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+        debug!(
+            "We have not seen the VID share for this view {:?} yet, so we cannot vote.",
+            proposal.view_number
+        );
+        return false;
+    };
+    let Some(vid_share) = vid_shares.get(&public_key).cloned() else {
+        debug!("we have not seen our VID share yet");
+        return false;
+    };
+
+    if let Some(upgrade_cert) = &decided_upgrade_cert {
+        if upgrade_cert.in_interim(cur_view)
+            && Some(proposal.block_header.payload_commitment())
+                != null_block::commitment(quorum_membership.total_nodes())
+        {
+            info!("Refusing to vote on proposal because it does not have a null commitment, and we are between versions. Expected:\n\n{:?}\n\nActual:{:?}", null_block::commitment(quorum_membership.total_nodes()), Some(proposal.block_header.payload_commitment()));
+            return false;
+        }
+    }
+
+    // Only vote if you have the DA cert
+    // ED Need to update the view number this is stored under?
+    let Some(cert) = consensus.saved_da_certs.get(&cur_view) else {
+        return false;
+    };
+
+    let view = cert.view_number;
+    // TODO: do some of this logic without the vote token check, only do that when voting.
+    let justify_qc = proposal.justify_qc.clone();
+    let parent = consensus
+        .saved_leaves
+        .get(&justify_qc.get_data().leaf_commit)
+        .cloned();
+
+    // Justify qc's leaf commitment is not the same as the parent's leaf commitment, but it should be (in this case)
+    let Some(parent) = parent else {
+        warn!(
+            "Proposal's parent missing from storage with commitment: {:?}, proposal view {:?}",
+            justify_qc.get_data().leaf_commit,
+            proposal.view_number,
+        );
+        return false;
+    };
+    let (Some(parent_state), _) = consensus.get_state_and_delta(parent.get_view_number()) else {
+        warn!("Parent state not found! Consensus internally inconsistent");
+        return false;
+    };
+    let Ok((validated_state, state_delta)) = parent_state
+        .validate_and_apply_header(
+            instance_state.as_ref(),
+            &parent,
+            &proposal.block_header.clone(),
+            vid_share.data.common.clone(),
+        )
+        .await
+    else {
+        warn!("Block header doesn't extend the proposal!");
+        return false;
+    };
+
+    let state = Arc::new(validated_state);
+    let delta = Arc::new(state_delta);
+    let parent_commitment = parent.commit();
+
+    let proposed_leaf = Leaf::from_quorum_proposal(&proposal);
+    if proposed_leaf.get_parent_commitment() != parent_commitment {
+        return false;
+    }
+
+    if let Err(e) = storage
+        .write()
+        .await
+        .update_undecided_state(
+            consensus.saved_leaves.clone(),
+            consensus.validated_state_map.clone(),
+        )
+        .await
+    {
+        warn!("Couldn't store undecided state.  Error: {:?}", e);
+    }
```

Comment:
Looks like we need to add this to the quorum vote task as well.

Reply:
We should probably just call this function and then use gates to call your task's specific code. I think that'll enable us to prevent sources of error from cropping up.

Reply:
Yeah, I think that's better. Then we don't have to make sure of consistent logic in multiple places.